Mathlib/GroupTheory/FreeGroup/CyclicallyReduced.lean

```diff
@@ -71,4 +72,79 @@ theorem flatten_replicate (n : ℕ) (h : IsCyclicallyReduced L) :
     exact h.2 _ ha _ hb
 
 end IsCyclicallyReduced
+
+/-- This function produces a subword of a word `w` by cancelling the first and last letters of `w`
+as long as possible. If `w` is reduced, the resulting word will be cyclically reduced. -/
+@[to_additive /-- This function produces a subword of a word `w` by cancelling the first and last
+letters of `w` as long as possible. If `w` is reduced, the resulting word will be cyclically
+reduced. -/]
+def reduceCyclically [DecidableEq α] : List (α × Bool) → List (α × Bool) :=
+  List.bidirectionalRec
+    (nil := [])
+    (singleton := fun x => [x])
+    (cons_append := fun a l b rC => if b.1 = a.1 ∧ (!b.2) = a.2 then rC else a :: l ++ [b])
+
+namespace reduceCyclically
+variable [DecidableEq α]
+
+/-- Partner function to `reduceCyclically`. See `reduceCyclically.conjugation`. -/
+@[to_additive /-- Partner function to `reduceCyclically`. See `reduceCyclically.conjugation`. -/]
+def conjugator : List (α × Bool) → List (α × Bool) :=
+  List.bidirectionalRec
+    (nil := [])
+    (singleton := fun _ => [])
+    (cons_append := fun a _ b rCC => if b.1 = a.1 ∧ (!b.2) = a.2 then a :: rCC else [] )
```

Comment:
Let's move `conjugator` and its results down until just above `conjugation`, so that we have both definitions with their definitional lemmas together.